runtime/src/main/java/org/corfudb/runtime/view/stream/IStreamView.java

```diff
@@ -0,0 +1,122 @@
+     * written at.
+     * @param   object
+     * @return  The (global) address the object was written at.
+     */
+    default long append(Object object) {
+        return append(object, null, null);
+    }
+
+    /** Retrieve the next entry from this stream, up to the tail of the stream
+     * If there are no entries present, this function will return NULL. If there
+     * are holes present in the log, they will be filled.
+     * @return  The next entry in the stream, or NULL, if no entries are
+     *          available.
+     */
+    default LogData next() {
+        return nextUpTo(Long.MAX_VALUE);
+    }
+
+    /** Retrieve the next entry from this stream, up to the address given or the
+     *  tail of the stream. If there are no entries present, this function
+     *  will return NULL. If there  are holes present in the log, they will
+     *  be filled.
+     * @param maxGlobal The maximum (global) address to read up to.
+     * @return          The next entry in the stream, or NULL, if no entries
+     *                  are available.
+     */
+    LogData nextUpTo(long maxGlobal);
+
+    /** Retrieve all of the entries from this stream, up to the tail of this
+     *  stream. If there are no entries present, this function
+     *  will return an empty list. If there  are holes present in the log,
+     *  they will be filled.
+     *
+     *  Note: the default implementation is thread-safe only if the
+     *  implementation of read is synchronized.
+     *
+     * @return          The next entries in the stream, or an empty list,
+     *                  if no entries are available.
+     */
+    default List<LogData> remaining() { return remainingUpTo(Long.MAX_VALUE); }
+
+    /** Retrieve all of the entries from this stream, up to the address given or
+     *  the tail of the stream. If there are no entries present, this function
+     *  will return an empty list. If there  are holes present in the log,
+     *  they will be filled.
+     *
+     *  Note: the default implementation is thread-safe only if the
+     *  implementation of read is synchronized.
+     *
+     * @param maxGlobal The maximum (global) address to read up to.
+     * @return          The next entries in the stream, or an empty list,
+     *                  if no entries are available.
+     */
+    default List<LogData> remainingUpTo(long maxGlobal) {
+        synchronized (this) {
+            final List<LogData> dataList = new ArrayList<>();
+            LogData thisData;
+            while ((thisData = nextUpTo(maxGlobal)) != null) {
+                dataList.add(thisData);
+            }
+            return dataList;
+        }
+    }
+
+    /** Returns whether or not there are potentially more entries in this
+     * stream - this function may return true even if there are not entries
+     * remaining, as addresses may have been acquired by other clients
+     * but not overwritten.
```

Comment:
but not overwritten -> but not written yet, or the addresses were hole-filled, or just failed.